server/src/main/java/org/apache/druid/query/RetryQueryRunner.java

```diff
@@ -0,0 +1,250 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiFunction;
+
+public class RetryQueryRunner<T> implements QueryRunner<T>
+{
+  private static final Logger LOG = new Logger(RetryQueryRunner.class);
+
+  private final QueryRunner<T> baseRunner;
+  private final BiFunction<Query<T>, List<SegmentDescriptor>, QueryRunner<T>> retryRunnerCreateFn;
+  private final RetryQueryRunnerConfig config;
+  private final ObjectMapper jsonMapper;
+
+  /**
+   * Runnable executed after the broker creates query distribution tree for the first attempt. This is only
+   * for testing and must not be used in production code.
+   */
+  private final Runnable runnableAfterFirstAttempt;
+
+  private int totalNumRetries;
+
+  public RetryQueryRunner(
+      QueryRunner<T> baseRunner,
+      BiFunction<Query<T>, List<SegmentDescriptor>, QueryRunner<T>> retryRunnerCreateFn,
+      RetryQueryRunnerConfig config,
+      ObjectMapper jsonMapper
+  )
+  {
+    this(baseRunner, retryRunnerCreateFn, config, jsonMapper, () -> {});
+  }
+
+  /**
+   * Constructor only for testing.
+   */
+  @VisibleForTesting
+  RetryQueryRunner(
+      QueryRunner<T> baseRunner,
+      BiFunction<Query<T>, List<SegmentDescriptor>, QueryRunner<T>> retryRunnerCreateFn,
+      RetryQueryRunnerConfig config,
+      ObjectMapper jsonMapper,
+      Runnable runnableAfterFirstAttempt
+  )
+  {
+    this.baseRunner = baseRunner;
+    this.retryRunnerCreateFn = retryRunnerCreateFn;
+    this.config = config;
+    this.jsonMapper = jsonMapper;
+    this.runnableAfterFirstAttempt = runnableAfterFirstAttempt;
+  }
+
+  @VisibleForTesting
+  int getTotalNumRetries()
+  {
+    return totalNumRetries;
+  }
+
+  @Override
+  public Sequence<T> run(final QueryPlus<T> queryPlus, final ResponseContext context)
+  {
+    return new YieldingSequenceBase<T>()
+    {
+      @Override
+      public <OutType> Yielder<OutType> toYielder(OutType initValue, YieldingAccumulator<OutType, T> accumulator)
+      {
+        final Sequence<Sequence<T>> retryingSequence = new BaseSequence<>(
+            new IteratorMaker<Sequence<T>, RetryingSequenceIterator>()
+            {
+              @Override
+              public RetryingSequenceIterator make()
+              {
+                return new RetryingSequenceIterator(queryPlus, context, baseRunner, runnableAfterFirstAttempt);
+              }
+
+              @Override
+              public void cleanup(RetryingSequenceIterator iterFromMake)
+              {
+                totalNumRetries = iterFromMake.retryCount;
+              }
+            }
+        );
+        return new MergeSequence<>(queryPlus.getQuery().getResultOrdering(), retryingSequence)
+            .toYielder(initValue, accumulator);
+      }
+    };
+  }
+
+  private List<SegmentDescriptor> getMissingSegments(QueryPlus<T> queryPlus, final ResponseContext context)
```

Comment:
Ah, this is a really good point, I agree it should be an option to be an exception if the list is truncated. 

I'm not sure what better we can really do without a fairly large refactoring involving making the broker keep track of which segments went to which direct druid client/sequence, and allowing historicals to basically give up if missing so many that the list would be truncated, returning an empty sequence and an indicator that **all** of the segments it was queried for should be retried.

Reply:
Hmm, I was aware of this issue, but ignored since the max size of the serialized responseContext is 7k. The missingSegments is a list of `SegmentDescriptor` which has an `interval`, a `version`, and a `partitionNumber`. Since its serialized size is about 80 bytes, each historical can report up to about 90 missing segments. I thought it wouldn't happen often, but well, I guess it depends on how your cluster is configured and how many segments the query reads.

> I'm not sure what better we can really do without a fairly large refactoring involving making the broker keep track of which segments went to which direct druid client/sequence, and allowing historicals to basically give up if missing so many that the list would be truncated, returning an empty sequence and an indicator that **all** of the segments it was queried for should be retried.

Hmm, I was thinking that we don't have to keep track of all segments, but can send a flag in the query context to historicals which makes the query failed immediately if the serialized responseContext is truncated. The broker will catch the failure and cancel the entire query. This won't retry on the truncated serialization, but I think it makes sense since it's a sort of failure of the retry system.

Reply:
Hmm, I was aware of this issue, but ignored since the max size of the serialized responseContext is 7k. The missingSegments is a list of `SegmentDescriptor` which has an `interval`, a `version`, and a `partitionNumber`. Since its serialized size is about 80 bytes, each historical can report up to about 90 missing segments. I thought it wouldn't happen often, but well, I guess it depends on how your cluster is configured and how many segments the query reads.

> I'm not sure what better we can really do without a fairly large refactoring involving making the broker keep track of which segments went to which direct druid client/sequence, and allowing historicals to basically give up if missing so many that the list would be truncated, returning an empty sequence and an indicator that **all** of the segments it was queried for should be retried.

Hmm, I was thinking that we don't have to keep track of all segments, but can send a flag in the query context to historicals which makes the query failed immediately if the serialized responseContext is truncated. The broker will catch the failure and cancel the entire query. This won't retry on the truncated serialization, but I think it makes sense since it's a sort of failure of the retry system.

Reply:
Added a flag and made the historicals fail when response context is truncated.